azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package azkaban.executor.container;
+
+import static azkaban.Constants.ContainerizedExecutionManagerProperties;
+
+import azkaban.executor.ExecutableFlow;
+import azkaban.executor.ExecutorLoader;
+import azkaban.executor.ExecutorManagerException;
+import azkaban.utils.Props;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This provides the ability to cleanup up execution containers which exceed the maximum allowed
+ * duration for a flow. Following extensions will be considered in future. (1) Convert this to a
+ * general background process for any other containerization related periodic tasks. (2) Add ability
+ * to track containers which may not be have been deleted despite prior attempts. (3) It's assumed
+ * final status updates to the db are handled by other parts, such as flow-executor. That can also
+ * be added here as a failsafe if needed.
+ */
+@Singleton
+public class ContainerCleanupManager {
+
+  private static final Logger logger = LoggerFactory.getLogger(ContainerCleanupManager.class);
+  private static final Duration DEFAULT_STALE_CONTAINER_CLEANUP_INTERVAL = Duration.ofMinutes(10);
+  private static final Duration DEFAULT_STALE_CONTAINER_AGE = Duration.ofDays(10);
+  private final long cleanupIntervalMin;
+  private final long staleContainerAgeDays;
+  private final ScheduledExecutorService cleanupService;
```

Comment:
Will you be adding service clean up as a part of this manager? You should create different thread for service and container. In fact different thread for any different types of clean up.

Reply:
I designed this to be agnostic to the specifics of the container implementation and the code here will interact only with the high level interface `ContainerImpl`. If a service needs to be deleted, that should be handled by the implementation of `ContainerImpl::deleteContainer()`. The K8S ContainerImpl does take care of it. 
For the same reason I don't think having a background thread for cleanup of every type of resource is necessary either. Please let me know if you think otherwise.

Reply:
Yes deleteContainer is taking care of deleting pod as well as service. 

Let's say flow has finished within 10 days then how will we terminate service for it?

Reply:
You're right, that needs to be supported directly within `deleteContainer` implementation. It will be added once the flow-status and healthcheck related changes are committed, as we discussed offline. I will create a ticket for it if it's not already tracked.